rust-runtime/aws-smithy-checksums/src/lib.rs

```diff
@@ -5,66 +5,192 @@
+        MD5_HEADER_NAME.clone()
+    } else {
+        HeaderName::from_static("x-amz-checksum-unknown")
+    }
+}
+
+/// Given a `HeaderName` representing a checksum algorithm, return the name of that algorithm
+/// as a `&'static str`.
+pub fn header_name_to_algorithm(checksum_header_name: &HeaderName) -> &'static str {
+    if checksum_header_name == CRC_32_HEADER_NAME {
+        CRC_32_NAME
+    } else if checksum_header_name == CRC_32_C_HEADER_NAME {
+        CRC_32_C_NAME
+    } else if checksum_header_name == SHA_1_HEADER_NAME {
+        SHA_1_NAME
+    } else if checksum_header_name == SHA_256_HEADER_NAME {
+        SHA_256_NAME
+    } else if checksum_header_name == MD5_HEADER_NAME {
+        MD5_NAME
+    } else {
+        "unknown-checksum-algorithm"
+    }
+}
+
+/// When a response has to be checksum-verified, we have to check possible headers until we find the
+/// header with the precalculated checksum. Because a service may send back multiple headers, we have
+/// to check them in order based on how fast each checksum is to calculate.
+pub const CHECKSUM_ALGORITHMS_IN_PRIORITY_ORDER: [&str; 4] =
+    [CRC_32_C_NAME, CRC_32_NAME, SHA_1_NAME, SHA_256_NAME];
 
 type BoxError = Box<dyn std::error::Error + Send + Sync>;
 
+/// Checksum algorithms are use to validate the integrity of data. Structs that implement this trait
+/// can be used as checksum calculators. This trait requires Send + Sync because these checksums are
+/// often used in a threaded context.
+pub trait Checksum: Send + Sync {
+    /// Given a slice of bytes, update this checksum's internal state.
+    fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError>;
+    /// Either return this checksum as a `HeaderMap` containing one HTTP header, or return an error
+    /// describing why checksum calculation failed.
+    fn headers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError>;
+    /// Return the `HeaderName` used to represent this checksum algorithm
+    fn header_name(&self) -> HeaderName;
+    /// "Finalize" this checksum, returning the calculated value as `Bytes` or an error that
+    /// occurred during checksum calculation. To print this value in a human-readable hexadecimal
+    /// format, you can print it using Rust's builtin [formatter].
+    ///
+    /// _**NOTE:** typically, "finalizing" a checksum in Rust will take ownership of the checksum
+    /// struct. In this method, we clone the checksum's state before finalizing because checksums
+    /// may be used in a situation where taking ownership is not possible._
+    ///
+    /// [formatter]: https://doc.rust-lang.org/std/fmt/trait.UpperHex.html
+    fn finalize(&self) -> Result<Bytes, BoxError>;
+    /// Return the size of this checksum algorithms resulting checksum, in bytes. For example, the
+    /// CRC32 checksum algorithm calculates a 32 bit checksum, so a CRC32 checksum struct
+    /// implementing this trait method would return 4.
+    fn size(&self) -> u64;
+}
```

Comment:
nice, I like this design! Have we already gone back and forth on this? I _might_ consider splitting out `HttpChecksum` as `CheckSum + headers/header_name